core/blockchain.go

```diff
@@ -2200,6 +2200,29 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 			return fmt.Errorf("Invalid new chain")
 		}
 	}
+	// Ensure XDPoS engine committed block will be not reverted
+	if xdpos, ok := bc.Engine().(*XDPoS.XDPoS); ok {
+		latestCommittedBlock := xdpos.EngineV2.GetLatestCommittedBlockInfo()
+		currentBlock := bc.CurrentBlock()
+		currentBlock.Number().Cmp(latestCommittedBlock.Number)
+		cmp := commonBlock.Number().Cmp(latestCommittedBlock.Number)
+		if cmp < 0 {
+			for _, oldBlock := range oldChain {
+				if oldBlock.Number().Cmp(latestCommittedBlock.Number) == 0 {
+					if oldBlock.Hash() != latestCommittedBlock.Hash {
+						log.Error("Impossible reorg, please file an issue", "oldnum", oldBlock.Number(), "oldhash", oldBlock.Hash(), "committed hash", latestCommittedBlock.Hash)
+					} else {
+						log.Warn("Stop reorg, blockchain is under forking attack", "old committed num", oldBlock.Number(), "old committed hash", oldBlock.Hash())
+						return fmt.Errorf("stop reorg, blockchain is under forking attack. old committed num %d, hash %x", oldBlock.Number(), oldBlock.Hash())
+					}
+				}
+			}
+		} else if cmp == 0 {
+			if commonBlock.Hash() != latestCommittedBlock.Hash {
+				log.Error("Impossible reorg, please file an issue", "oldnum", commonBlock.Number(), "oldhash", commonBlock.Hash(), "committed hash", latestCommittedBlock.Hash)
```

Comment:
return here?

Reply:
I see below there is Ethereum's code without return. This code should never happen, yet not return may help the program continue to run and find out more bugs.
```
log.Error("Impossible reorg, please file an issue", "oldnum", oldBlock.Number(), "oldhash", oldBlock.Hash(), "newnum", newBlock.Number(), "newhash", newBlock.Hash())
```

Reply:
But if you don't return from here, it may possibly reorg if there is no other error occurs

Reply:
In this condition, I think we can let reorg run. For example, three blocks at the same number, block A (committed) and B, C. Reorg from B to C doesn't matter since A is committed.

Reply:
ok cool!